include/ginkgo/core/preconditioner/batch_jacobi.hpp

```diff
@@ -0,0 +1,388 @@
+               this->get_block_offset(block_id, block_storage_cumulative);
+    }
+
+    /**
+     * Returns the stride between the rows of the block.
+     *
+     * @param block_idx  the id of the block from the perspective of individual
+     * batch entry
+     * @param block_ptrs  the block pointers array
+     *
+     * @return stride between rows of the block
+     */
+    GKO_ATTRIBUTES size_type get_stride(
+        const int block_idx, const IndexType* const block_ptrs) const noexcept
+    {
+        return block_ptrs[block_idx + 1] - block_ptrs[block_idx];
+    }
+};
+
+
+/**
+ * A block-Jacobi preconditioner is a block-diagonal linear operator, obtained
+ * by inverting the diagonal blocks (stored in a dense row major fashion) of the
+ * source operator.
+ *
+ * With the batched preconditioners, it is required that all items in the batch
+ * have the same sparsity pattern. The detection of the blocks and the block
+ * pointers require that the sparsity pattern of all the items be the same.
+ * Other cases is undefined behaviour. The input batch matrix must be in
+ * batch::Csr matrix format or must be convertible to batch::Csr matrix format.
+ * The block detection algorithm and the conversion to dense blocks kernels
+ * require this assumption.
+ *
+ * @note In a fashion similar to the non-batched Jacobi preconditioner, the
+ * maximum possible size of the diagonal blocks is equal to the maximum warp
+ * size on the device (32 for NVIDIA GPUs, 64 for AMD GPUs).
+ *
+ * @tparam ValueType  value precision of matrix elements
+ * @tparam IndexType  index precision of matrix elements
+ *
+ * @ingroup jacobi
+ * @ingroup precond
+ * @ingroup BatchLinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Jacobi : public EnableBatchLinOp<Jacobi<ValueType, IndexType>> {
+    friend class EnableBatchLinOp<Jacobi>;
+    friend class EnablePolymorphicObject<Jacobi, BatchLinOp>;
+
+public:
+    using EnableBatchLinOp<Jacobi>::convert_to;
+    using EnableBatchLinOp<Jacobi>::move_to;
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = batch::matrix::Csr<ValueType, IndexType>;
+
+    /**
+     * Returns the storage scheme used for storing Batched Jacobi blocks.
+     *
+     * @return the storage scheme used for storing Batched Jacobi blocks
+     *
+     */
+    const batched_jacobi_blocks_storage_scheme<index_type>&
+    get_blocks_storage_scheme() const noexcept
+    {
+        return blocks_storage_scheme_;
+    }
+
+    /**
+     *  Returns the block pointers.
+     *  @note Returns nullptr in case of a scalar jacobi preconditioner
+     * (max_block_size = 1).
+     *  @return the block pointers
+     */
+    const index_type* get_const_block_pointers() const noexcept
+    {
+        if (parameters_.max_block_size == 1) {
+            return nullptr;
+        }
+        return parameters_.block_pointers.get_const_data();
```

Comment:
```suggestion
        return parameters_.block_pointers.get_const_data();
```
Any empty array will return nullptr, so no need to distinguish the two cases. Same for the other functions